Update opcua library

JGrothoff opened the pull request:
Simplified opcua server library and updated to open62541 version v1.3.

### ToDos
- [x] @mhthies: Include open62541 as submodule and compile via cmake 
- [x] Write README.md with basic usage and development tutorial for custom transformations
  - [x] Document how to compile library, e.g. how to set Minimal/Reduced/Full NS0
  - [x] Document design decisions and outlook
- [x] Add example project
  - [x] Add simple example for custom interface, e.g. similiar to ipsms
  - [x] Add template for custom interface
  - [x] Explain or show prototypical: how to use OPC UA Companion Specifications
- [x] Add functionality to use custom datatypes and load nodesets, that were parsed via open62541 parser

JGrothoff committed:
Merged and simplified opcua library

* Interfaces
  * Delete `store`, `datatype` and `depencies` variables
  * Remove `nodeset`, `unload` function
  * Make `load` function abstract
  * Adapt opcua_ovInterface.c
  * Allow to switch off ovInterface -> Documentation: Overwrite CheckNode function in CustomInterface with 'true'
* ServerToInterface
  * Allow only one custom interface: OneToOne
  * Update `link` and `unlink` to forbid linking while server is running and don't load or unload dependent interfaces
* Server
  * Move serverConfig variables to uaServer

Co-authored-by: TorbenD <t.miny@plt.rwth-aachen.de>

JGrothoff committed:
Fixed missing return in ovSwitch.

Co-authored-by: TorbenD <t.miny@plt.rwth-aachen.de>

JGrothoff committed:
Updated to open62541 to v1.2.3

Added namespace based nodestore switch (nsSwitch).

Co-authored-by: TorbenD <t.miny@plt.rwth-aachen.de>

JGrothoff committed:
Minor modifications to checkNode and checkReference

co-authored-by: TorbenD <t.miny@plt.rwth-aachen.de>

JGrothoff committed:
Updated load, checkReference and checkNode function signatures.

co-authored-by: TorbenD <t.miny@plt.rwth-aachen.de>

JGrothoff commented:
Added example project: https://github.com/acplt/rte_opcua_example

JGrothoff committed:
Removed unecessary abstract function definitions.

JGrothoff committed:
Added parameter for addReferences, for browsenames from other namespaces.

JGrothoff committed:
Reverted deletion of necessary abstract functions.

JGrothoff committed:
Removed direction definitions as they are not needed by the library.

JGrothoff committed:
Added some first documentation on opcua library.

JGrothoff committed:
Added open62541 as submodule.

JGrothoff committed:
Compile with open62541 submodule instead.

Moved from v1.2.3 tag to open62541 branch 1.2 to use shallow checkout of submodule. Adapted to new getNode and getNodePtr function accordingly.

JGrothoff committed:
Bugfix allow other libraries to use open62541

mhthies commented on syslibs/opcua/include/helpers.h:
```diff
@@ -0,0 +1,59 @@
+/*
+ * helpers.h
```
The `include/` directory of OV libs is added as header search path *publicly* (see [ov_functions.cmake:27](https://github.com/acplt/rte/blob/master/cmake/ov_functions.cmake#L27)), i.e. other libraries depending on the opcua library will also search for header files in opcua's include directory. Thus, `opcua_helpers.h` was actually the better file name, since it will less likely result in naming collisions.

JGrothoff replied:
So, i guess we should add the library prefix for other headers too?
* nsSwitch.h
* ovSwitch.h
* ovTrafo.h

JGrothoff replied:
Should we keep the source file names without library prefix?

mhthies replied:
For the source files it should not make any difference. However, it is a nice convention to have the corresponding .h and .c files' names match to make it easier to explore the code.

mhthies commented on syslibs/opcua/CMakeLists.txt:
```diff
@@ -1,15 +1,22 @@
     target_link_libraries(opcua PRIVATE winmm.lib ws2_32.lib)
 endif()
+
+target_link_libraries(opcua PUBLIC open62541)
```
Should the library actually be linked *PUBLIC*ly? I.e. shall all dependent libraries/executables also directly be linked with open62541?

JGrothoff replied:
Nope. I will change that back to private. (I thought this is needed for the example project, but instead the target link library is should be added there.)

mhthies replied:
Oh, well, I see from the CI why the public linking was required: https://github.com/acplt/rte/runs/4941930184?check_suite_focus=true#step:4:672
Since open62541 types are used in the library's .ovt file, which *is* publicly shared with all dependant OV libraries, the open62541 header files (and library) need to be as well.

mhthies commented on .gitmodules:
```diff
@@ -0,0 +1,5 @@
+	path = syslibs/opcua/open62541
+	url = https://github.com/open62541/open62541
+	shallow = true
+	branch = 1.2
```
Don't you want to use the `pack/1.2` branch of open62541, which does not require recursive submodules?

JGrothoff replied:
The current version does not require the recursive submodules (`deps/mdnsd` and `deps/ua-nodeset`). Hence, we don't need to stick with the *pack* versions.

I noticed, that the combination of *branch* and *shallow* are not a good idea for submodules (c.f. [shallow submodules](https://www.reddit.com/r/git/comments/5b8s8w/git_shallow_submodules/)) and that the current commit referenced by the submodule doesn't even belong to the 1.2 branch. (It's a commit from the open62541 master - my bad). I will downgrade to latest stable commit from branch 1.3. As mentioned in the [release candidate for 1.3](https://github.com/open62541/open62541/releases/tag/v1.3-rc1) the api is frozen for 1.3. So we should be fine with any future version of 1.3 (so that we can add the find_package CMake option with version 1.3).

Although, the [current changes to the nodestore api](https://github.com/open62541/open62541/commit/6b8db940e5fb4699c7bcde777fc7b21234cc947b) seem promising for future optimizations of the transformation concept. (i will add that to the library readme for future generations 😃)

mhthies commented on syslibs/opcua/CMakeLists.txt:
```diff
@@ -1,15 +1,22 @@
+set(UA_AMALGAMATION_ARCHITECTURES "posix;win32" CACHE STRING "Include posix and win32 architecture." FORCE)
+set(UA_ENABLE_IMMUTABLE_NODES "ON" CACHE BOOL "Use getNodeCopy and replaceNode instead of modifying nodes directly." FORCE)
+set(UA_ENABLE_NODEMANAGEMENT "OFF" CACHE BOOL "Nodemanagement is currently not implemented, as no backtransformations are implemented." FORCE)
+add_subdirectory(open62541)
```
A nice feature here would be defining a custom CMake option to switch between including the open62541 sources from the subdirectory and linking to a system-provided pre-build open62541 by using `find_package(open62541 ...)` as explained in https://open62541.org/doc/current/installing.html#manual-installation.

Additionally, we should make sure, open62541 is only built/searched when required. We could add an additional `OV_WITH_OPCUA` option, and bascially wrap this full CMakeLists file into an `if(OV_WITH_OPCUA) … endif()`. When disabling this option, open62541 will not be included/searched, but the opcua OV library will not be available as well.

mhthies replied:
> I don't see the need for an OV_WITH_OPCUA option, as the library can be switched ov via SET(BUILD_OV_LIBRARY_opcua OFF CACHE BOOL "") analogous to kshist if desired.

However, if you set `BUILD_OV_LIBRARY_opcua` to off, open62541 will still be included into the CMake project. This may result in unnecessarily building open62541 and may lead to further problems when open62541 is not available (neither as submodule nor as a systemwide-installation). It may also result in problems when building OV for a platform that is not not supported by open62541. In those situations, it would be nice to have a switch, that disables including open62541 all together.

JGrothoff replied:
Ok, so can we just reuse the `BUILD_OV_LIBRARY_opcua` option, as shown in 6bab542.

mhthies replied:
Hm. Originally, the `BUILD_OV_LIBRARY_*` has other semantics: It controls whether the library is added to CMake's default build target, i.e. it is build when simply running `make` (or whatever build system you are using) without a specific target. When you disable the default build using the relevant option, it should (by now) still be possible to build the library explicitly (`make opcua`) or as a dependency of another library.

This is why I propose to add another flag (something like `OV_WITH_OPCUA`) to completely disable opcua/open62541 support.

mhthies commented on syslibs/opcua/CMakeLists.txt:
```diff
@@ -1,15 +1,22 @@
+#alternative syntax:
+set(UA_NAMESPACE_ZERO "REDUCED" CACHE STRING "Enable reduced NS0 (default) to allow for faster server startup." FORCE)
+set(UA_ENABLE_AMALGAMATION "ON" CACHE BOOL "Merge als .c and .h files from open62541 in one open62541.c/.h" FORCE)
+set(UA_AMALGAMATION_ARCHITECTURES "posix;win32" CACHE STRING "Include posix and win32 architecture." FORCE)
```
Why are the AMALGAMATION options required?

JGrothoff replied:
I guess, its just for lazy including without the need for searching the right headers. I will change that.

mhthies reviewed (commented):
I didn't really review all the C code (I assume, you know what you are doing with OV and open62541 :) ), but I still have some comments, esp. about project structure things and from an OV user perspective.

JGrothoff committed:
Rollback to open62541 branch 1.3

JGrothoff committed:
Revert public linking of open62541.

JGrothoff committed:
Added submodule checkout for github actions.

HolgerJeromin commented on syslibs/opcua/source/helpers.c:
```diff
@@ -797,27 +784,29 @@ OV_DLLFNCEXPORT UA_Int32 opcua_nodeStoreFunctions_getVtblPointerAndCheckAccess(O
+	if(pelem->elemtype == OV_ET_VARIABLE) //FIXME Allow read access for all variables, even if they don't have get/set accesors
+		*access = OV_AC_READWRITE;
+	else
+		*access = (*ppVtblObj)->m_getaccess(pelem->pobj, pelem, NULL);
```
See cshmi for an advanced getaccess example

https://github.com/acplt/rte/blob/e34398cbb7ea86ed693cb9a0db209c4ca033f18a/syslibs/cshmi/source/baseclass_Object.c#L180

JGrothoff replied:
Thanks. I changed the code to use the [ov_object_getAccessEx](https://github.com/acplt/rte/blob/cd6044bd14358a7915389b70116c7eeb1cb20474/core/ov/source/ov_object.c#L521-L551) instead, which is slightly different from cshmi_Object_getaccess: getAccessEx gives read access to `OV_VT_POINTER`. Both implementations seem to give read access to `OV_VT_STRUCT`. Maybe we should align this.

JGrothoff committed:
Modified logger options.

JGrothoff committed:
Added option to use find_package for open62541

JGrothoff committed:
Reverted again to PUBLIC linking of open62541

"Since open62541 types are used in the library's .ovt file, which is publicly shared with all dependant OV libraries, the open62541 header files (and library) need to be as well." (mhthies)

JGrothoff committed:
Deleted WIN32 undefs for open62541 in ovt

JGrothoff committed:
Moved application name and uri defines to helpers

JGrothoff committed:
Reworked comment on useOvTrafo variable.

JGrothoff committed:
Added opcua prefix for header files.

"The include/ directory of OV libs is added as header search path publicly (see ov_functions.cmake:27), i.e. other libraries depending on the opcua library will also search for header files in opcua's include directory." (mhthies)

JGrothoff committed:
Removed forcing of amalagamtion for open62541

JGrothoff committed:
Fixed warning for unused variables.

JGrothoff committed:
Fix missing include for nsSwitch.

JGrothoff committed:
Only include open62541 if opcua should be build.

JGrothoff committed:
Try to fix MSVC errors with WINSOCKAPI define

JGrothoff commented on syslibs/opcua/model/opcua.ovt:
```diff
@@ -1,8 +1,10 @@
+*/
 #ifdef WIN32
 #undef WIN32
 #endif
```
[This problem is caused when including <windows.h> before <winsock2.h>](https://stackoverflow.com/questions/1372480/c-redefinition-header-files-winsock2-h), which seems to be the problem. But windows.h and winsock2.h are included in varios places distributed over ov and open62541.

bjoernhoeper replied:
Hey @JGrothoff. I do not have a direct solution but please take a look at the `WIN32_LEAN_AND_MEAN` define. When using this in the right place usually resolves the header problem (Further reference: https://devblogs.microsoft.com/oldnewthing/20091130-00/?p=15863)

JGrothoff replied:
Thanks, but defining `PLT_USE_WINSOCK2` did the [trick](https://github.com/acplt/rte/blob/master/core/runtimeserver/ov_ks_missing_symbols.h#L64-L69). I guess, this [prevents ov from including winsock.h](https://github.com/acplt/rte/blob/master/core/ovheaders/ov_config.h#L263-L267), which collides with winsock2.h.

JGrothoff committed:
Try to fix MSVC errors with WIN32_LEAN_AND_MEAN or PLT_USE_WINSOCK2

JGrothoff committed:
Fix MSVC errors for winsock2 and missing dllexport

JGrothoff committed:
Renamed source files: added library prefix to match headers

JGrothoff committed:
Add option to completly disable opcua library.

JGrothoff committed:
Changed getAccess to use ov_object_getAccessEx

JGrothoff committed:
Elaborate on interfaces and open62541 defines

JGrothoff committed:
Add some documentation on remarks.
